### PR TITLE
Alternative for: Filter update operations during resharding

### PR DESCRIPTION
**Pull request by timvisee**
Alternative resharding pre filter implementation on top of the current PR (<https://github.com/qdrant/qdrant/pull/4867>).

Reimplements the resharding pre filter.

This functions more as a request for comment, rather than a direct replacement.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### CODE REVIEW

File: lib/collection/src/collection/resharding_update_pre_filter.rs [hunk 1]
```diff
@@ -62,32 +65,33 @@ impl ReshardingUpdatePreFilter {
...
-        // - on *receiver* shards during `MigratingPoints` stage
-        // - and on *sender* shards during `ReadHashRingCommitted` stage when resharding *up*
-        let should_filter = (is_receiver_shard && state.stage == ReshardStage::MigratingPoints)
-            || (is_sender_shard
-                && state.stage >= ReshardStage::ReadHashRingCommitted
-                && state.direction == ReshardingDirection::Up);
-
-        if !should_filter {
+        // After write hash ring commit, we stop filtering
+        if state.stage >= ReshardStage::WriteHashRingCommitted {
```
Comment (ffuugoo) [inline lib/collection/src/collection/resharding_update_pre_filter.rs:69]: I never explained this explicitly, but I don't think we can do this. E.g.:

- Node A already received `WriteHashRingCommitted` message from consensus
  - and so it does not apply pre-filter locally anymore
- but Node B did not yet received `WriteHashRingCommitted` from consensus
  - and so it will still send all operations to new and *old* shards both
- now, user sends `upsert point 1337` to Node B
  - Node B forwards this update to *old* shard on Node A
  - and Node A inserts this point into *old* shard, because it does not apply pre-filter anymore

Basically, I think we should apply pre-filter until resharding is finished.

**Comment by ffuugoo** [comment]
Tbh, I don't see much difference with #4867.

File: lib/collection/src/collection/resharding_update_pre_filter.rs (hunk 1 above)
Comment (timvisee) [reply lib/collection/src/collection/resharding_update_pre_filter.rs:69]: > Basically, I think we should apply pre-filter until resharding is finished.

Fine by me. We can eventually fine tune it if desired.

**Comment by timvisee** [comment]
> Tbh, I don't see much difference with #4867.

I agree it's not very different. I'd like to have seen a more significant change.

But I do see this:
- easier to read apply function
- slightly different decision tree better representing what we want

**Comment by ffuugoo** [comment]
Closing, cause we implemented a different approach in https://github.com/qdrant/qdrant/pull/4928.
